File: javaScript/arrays&hashing/productOfArrayExceptSelf.js
/*
Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements 
of nums except nums[i].
The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
You must write an algorithm that runs in O(n) time and without using the division operation.
*/
/**
 * @param {number[]} nums
 * @return {number[]}
 */
const productExceptSelf = function (nums) {
  let start = 1;
  let res = [];
  for (let i = 0; i < nums.length; i++) {
    res.push(start);
    start *= nums[i];
    //[1,1,2,6]
  }
  let start2 = 1;
  for (let i = nums.length - 1; i >= 0; i--) {
    //i starts from nums[3] = 4
    //starting from the last element of nums
    //multiply the current element of res with start2
    //then multiply start2 with the current element of nums
    //[24,12,4,1]
    res[i] *= start2;
    start2 *= nums[i];
  }
  return res;
};

console.log(productExceptSelf([1, 2, 3, 4])); //[24,12,8,6]

File: python/educative.io/lists/productOfArrayExceptSelf.py
""" 
You're given an integer array, nums. Return a resultant array product so that product[i] is 
equal to the product of all the elements of nums except nums[i].

Write an algorithm that runs in O(n)
time without using the division operation.
"""

from typing import List
def product_except_self(nums:List[int]) -> List[int]:
    left = 1
    ans = [0] * len(nums)
    for i in range(len(nums)):
        ans[i] = left
        left *= nums[i]
    
    print(ans) # [1,1,2,6]
    
    right = 1
    for i in range(len(nums) -1, -1, -1):
        ans[i] = right * ans[i]
        right = right * nums[i]
    
    return ans

print(product_except_self([1,2,3,4]))